app/js/MigrationPersistor.js

```diff
@@ -68,46 +69,158 @@ module.exports = function(primary, fallback) {
     }
   }
 
-  async function _copyFileFromFallback(
+  async function _getFileStreamAndCopyIfRequired(bucketName, key, opts) {
+    const shouldCopy =
+      Settings.filestore.fallback.copyOnMiss && !opts.start && !opts.end
+
+    try {
+      return await primary.promises.getFileStream(bucketName, key, opts)
+    } catch (err) {
+      if (err instanceof NotFoundError) {
+        const fallbackBucket = _getFallbackBucket(bucketName)
+        if (shouldCopy) {
+          return _copyFileFromFallback(
+            fallbackBucket,
+            bucketName,
+            key,
+            key,
+            true
+          )
+        } else {
+          return fallback.promises.getFileStream(fallbackBucket, key, opts)
+        }
+      }
+      throw err
+    }
+  }
+
+  async function _copyFromFallbackStreamAndVerify(
+    stream,
     sourceBucket,
     destBucket,
     sourceKey,
     destKey
   ) {
+    try {
+      let sourceMd5
+      try {
+        sourceMd5 = await fallback.promises.getFileMd5Hash(
+          sourceBucket,
+          sourceKey
+        )
+      } catch (err) {
+        logger.warn(err, 'error getting md5 hash from fallback persistor')
+      }
+
+      await primary.promises.sendStream(destBucket, destKey, stream, sourceMd5)
+    } catch (err) {
+      let error = err
+      metrics.inc('fallback.copy.failure')
+
+      try {
+        await primary.promises.deleteFile(destBucket, destKey)
+      } catch (err) {
+        error = new WriteError({
```

Comment:
I'm trying to avoid unnecessary log lines if possible. The `withCause` will wrap the error, and the error message made it clear what went wrong here so I think this is OK. It gives us a little extra context without having to resort to another `logger.warn` at this point.

Reply:
To be more specific, the error catched from line 115 is lost.

The error message seen outside `_copyFromFallbackStreamAndVerify` may be simplified as
```
unable to copy file to destination persistor
withCause: unable to clean up destination copy artifact
withCause: some PRIMARY.deleteFile error
```
But this is more important:

```
unable to copy file to destination persistor
withCause: some PRIMARY.sendStream error
withCause: ...
```

Reply:
Oh, yes I see what you mean.

Reply:
Swizzled things around to make the cleanup error a property of the main error. :+1: